# RB Model

In [1]:
# imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sys
import os

# models
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor, StackingRegressor
from xgboost import XGBRegressor
from sklearn.pipeline import Pipeline

# data preprocessing, performance metrics
from sklearn.preprocessing import RobustScaler
from sklearn.model_selection import train_test_split, cross_validate, cross_val_predict
from sklearn.metrics import r2_score, mean_squared_error

# helper functions
# sys.path.append(r'C:\Users\heefj\OneDrive\Documents\nfl\fantasy_2024\notebooks\models.py')
# sys.path.append(r'C:\Users\heefj\OneDrive\Documents\nfl\fantasy_2024\notebooks\plotting.py')
from models import create_features, cross_val
from plotting import plot_mean_and_counts, plot_ranks_line, plot_ranks_boxplot

# display
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', 200)

# global random_state
random_state = 9

In [2]:
# load data
df = pd.read_csv('../data/final_data/master.csv')

# get RBs
rb = df[df['Pos'] == 'RB']

# view
print(f'The RB data has {rb.shape[0]} rows and {rb.shape[1]} columns:')
rb.head()

The RB data has 9297 rows and 115 columns:


Player   Tm Pos  Age   G  GS  Pass_Cmp  Pass_Att  Pass_Yds  Pass_TD  \
0    Ron Johnson  NYG  RB   23  14  14         0         0         0        0   
5     Dave Smith  SDG  RB   23   7   1         0         0         0        0   
6      Nick Eddy  DET  RB   26  11   0         0         0         0        0   
7  Craig Baynham  CHI  RB   26   5   5         0         0         0        0   
8  Chuck Mercein  NYJ  RB   27   9   0         0         0         0        0   

   Pass_Int  Rush_Att  Rush_Yds  Rush_Y/A  Rush_TD    Rec_Tgt  Rec_Rec  \
0         0       263      1027  3.904943        8  91.185521       48   
5         0        14        42  3.000000        0   7.598793        4   
6         0        18        47  2.611111        1   7.598793        4   
7         0        26        68  2.615385        0  22.796380       12   
8         0        20        44  2.200000        0   5.699095        3   

   Rec_Yds    Rec_Y/R  Rec_TD  Fmb  FmbLost       Key  Year  Scrim_Yds  \
0      487  10.145833       4  5.0     2.75  JohnRo02  1970       1514   
5       65  16.250000       0  0.0     0.00  SmitDa01  1970        107   
6       22   5.500000       0  2.0     1.10  EddyNi00  1970         69   
7       43   3.583333       0  0.0     0.00  BaynCr00  1970        111   
8       27   9.000000       1  1.0     0.55  MercCh00  1970         71   

   Scrim_TD  num_games  games_played_pct  games_started_pct  ProBowl  AllPro  \
0        12         16            0.8750           1.000000        1       1   
5         0         16            0.4375           0.142857        0       0   
6         1         16            0.6875           0.000000        0       0   
7         0         16            0.3125           1.000000        0       0   
8         1         16            0.5625           0.000000        0       0   

   Exp  New_Team  Will_be_on_New_Team  Pass_Y/A  Cmp%    Catch%  Touches  \
0    1         0                    0       0.0   0.0  0.526399      311   
5    1         0                    0       0.0   0.0  0.526399       18   
6    4         0                    0       0.0   0.0  0.526399       22   
7    4         0                    0       0.0   0.0  0.526399       38   
8    5         0                    0       0.0   0.0  0.526399       23   

   Pass_Cmp_per_game  Pass_Att_per_game  Pass_Yds_per_game  Pass_TD_per_game  \
0                0.0                0.0                0.0               0.0   
5                0.0                0.0                0.0               0.0   
6                0.0                0.0                0.0               0.0   
7                0.0                0.0                0.0               0.0   
8                0.0                0.0                0.0               0.0   

   Pass_Int_per_game  Rush_Att_per_game  Rush_Yds_per_game  Rush_TD_per_game  \
0                0.0          18.785714          73.357143          0.571429   
5                0.0           2.000000           6.000000          0.000000   
6                0.0           1.636364           4.272727          0.090909   
7                0.0           5.200000          13.600000          0.000000   
8                0.0           2.222222           4.888889          0.000000   

   Rec_Tgt_per_game  Rec_Rec_per_game  Rec_Yds_per_game  Rec_TD_per_game  \
0          6.513251          3.428571         34.785714         0.285714   
5          1.085542          0.571429          9.285714         0.000000   
6          0.690799          0.363636          2.000000         0.000000   
7          4.559276          2.400000          8.600000         0.000000   
8          0.633233          0.333333          3.000000         0.111111   

   Fmb_per_game  FmbLost_per_game  Scrim_Yds_per_game  Scrim_TD_per_game  \
0      0.357143          0.196429          108.142857           0.857143   
5      0.000000          0.000000           15.285714           0.000000   
6      0.181818          0.100000            6.2

Many of the 9000 RB seasons that we have are seasons of backup players with very little volume. The goal of this model is to predict the performance of fantasy-relevant players in 2024. We will identify all RBs who have finished at or above replacement-level (top 25) at least 1 time in their career. This will be the subset that our model will train on.

In [3]:
# get all unique 'Key' values that have at least 1 'VORP_ppr_10tm' value >= 0
keys = rb[rb['VORP_ppr_10tm'] >= 0]['Key'].unique()

# get all unique Keys from 2023
keys_2023 = df[df['Year'] == 2023]['Key'].unique()

# drop all rows with 'Key' values not in either key set
rb = rb[(rb['Key'].isin(keys)) | (rb['Key'].isin(keys_2023))]

# view
print(f'The new RB data has {rb.shape[0]} rows.')

The new RB data has 3834 rows.


Now, we have 3834 seasons of RBs who were relevant for at least one season in their career.

## Create Features

In [4]:
# get 2023 RBs
rb_2023 = rb[rb['Year'] == 2023]

In [5]:
# define column lists
meta_cols = ['Key', 'Age', 'Exp']
volume_cols = ['Rush_Att_per_game', 'Rec_Tgt_per_game', 'Rec_Rec_per_game']
production_cols = ['Rush_Yds_per_game', 'Rush_TD_per_game', 'Rec_Yds_per_game', 'Rec_TD_per_game']
health_cols = ['games_played_pct', 'games_started_pct']
rank_col = ['PPGPosRank_ppr']

# combine and add target
feature_subset = meta_cols + volume_cols + production_cols + health_cols# + rank_col

# define target
target = 'PPGTarget_ppr'

# get subset
rb = rb[feature_subset + [target]]
rb_2023 = rb_2023[feature_subset + [target]]

# drop nulls 
rb = rb.dropna()

# look at shapes
print(f'2023 RBs: {rb_2023.shape[0]} rows')
print(f'RB seasons to train on: {rb.shape[0]} rows')

2023 RBs: 166 rows
RB seasons to train on: 3247 rows


In [6]:
# create features
features_rb = create_features(rb)

# drop Key
features_rb = features_rb.drop(columns='Key')

## Model

In [7]:
# view std of target
print(f'The std of the target is {features_rb[target].std()}')

The std of the target is 5.772531420035651


When evaluating RMSE (the average distance the model's prediction is from the true target), I will aim to get it below __5.77__ (the standard deviation of the target).

In [8]:
# load models_df
models_df = pd.read_csv('../data/models/models_df.csv')

In [9]:
# define some base models
lr = LinearRegression()
rf = RandomForestRegressor()
models = [lr, rf]

# cross validate the models
for model in models:
    cross_val(df=features_rb, pos='RB', target=target, estimator=model, models_df=models_df)

# save models_df
models_df.to_csv('../data/models/models_df.csv', index=False)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 24 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:    1.7s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 24 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   17.5s finished


In [10]:
# view top 3 models
models_df#.sort_values('Mean_RMSE', ascending=True).head(3)

Pos                                           Features         Target  \
0   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
1   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
2   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
3   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
4   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
5   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
6   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
7   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
8   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
9   RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
10  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
11  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
12  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
13  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
14  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
15  RB  Index(['Age', 'Exp', 'Will_be_on_New_Team', 'P...  PPGTarget_ppr   
16  RB  Index(['Age', 'Exp', 'Rush_Att_per_game', 'Rec...  PPGTarget_ppr   
17  RB  Index(['Age', 'Exp', 'Rush_Att_per_game', 'Rec...  PPGTarget_ppr   
18  RB  Index(['Age', 'Exp', 'Rush_Att_per_game', 'Rec...  PPGTarget_ppr   
19  RB  Index(['Age', 'Exp', 'Rush_Att_per_game', 'Rec...  PPGTarget_ppr   

                                                Model  Mean_RMSE   Mean_R2  
0                                  LinearRegression()   4.430300  0.385417  
1                               KNeighborsRegressor()   4.816332  0.272903  
2                             RandomForestRegressor()   4.474463  0.372951  
3   XGBRegressor(base_score=None, booster=None, ca...   4.801177  0.278411  
4                                  LinearRegression()   4.362126  0.401467  
5                               KNeighborsRegressor()   4.881338  0.253541  
6                             RandomForestRegressor()   4.468689  0.375771  
7   XGBRegressor(base_score=None, booster=None, ca...   4.703529  0.307937  
8                                  LinearRegression()   4.352444  0.404066  
9                               KNeighborsRegressor()   4.880525  0.255243  
10                            RandomForestRegressor()   4.470724  0.374936  
11  XGBRegressor(base_score=None, booster=None, ca...   4.688887  0.311126  
12                                 LinearRegression()   4.329753  0.410262  
13                            RandomForestRegressor()   4.472852  0.374629  
14                                 LinearRegression()   4.342244  0.406731  
15                            RandomForestRegressor()   4.450841  0.380248  
16                                 LinearRegression()   4.358915  0.402069  
17                            RandomForestRegressor()   4.468755  0.375774  
18                                 LinearRegression()   4.377656  0.396998  
19                            RandomForestRegressor()   4.478966  0.372792

## 2024 Predictions
Here, we will predict the PPG for RBs in the 2024 season.

In [22]:
# drop target col from 2023 players
rb_2023 = rb_2023.drop(columns=[target])

# create features for 2023 RBs, drop Key
features_2023 = create_features(rb_2023).drop(columns='Key')

In [23]:
# best model
model = LinearRegression()

# training features and target
X_train = features_rb[feature_subset[1:]]
y_train = features_rb[target]

# 2023 rows are test set
X_test = features_2023[feature_subset[1:]]

# create pieline
pipeline = Pipeline([
        ('scaler', RobustScaler()),
        ('model', model)])

# train on entire dataset
pipeline.fit(X_train, y_train)

# predict
preds = pipeline.predict(X_test)

# add preds to 2023 df
rb_2023['2024_pred_PPG_ppr'] = preds

In [24]:
# create dict mapping Key to Player name
key_to_player = dict(zip(df['Key'], df['Player']))

# map Key to Player name
rb_2023['Player'] = rb_2023['Key'].map(key_to_player)

# sort and add "2024_pred_PosRank_ppr"
rb_2023 = rb_2023.sort_values(by='2024_pred_PPG_ppr', ascending=False).reset_index(drop=True)
rb_2023['2024_pred_PosRank_ppr'] = rb_2023.index + 1

# view top 25
rb_2023[['Player', '2024_pred_PPG_ppr', '2024_pred_PosRank_ppr']].head(25).T

0               1             2   \
Player                 Christian McCaffrey  Kyren Williams  Jahmyr Gibbs   
2024_pred_PPG_ppr                17.162275       16.333419     15.330918   
2024_pred_PosRank_ppr                    1               2             3   

                                  3            4               5   \
Player                 De'Von Achane  Breece Hall  Bijan Robinson   
2024_pred_PPG_ppr            15.1789    15.158894       14.098164   
2024_pred_PosRank_ppr              4            5               6   

                                  6               7                8   \
Player                 Rachaad White  Travis Etienne  Jonathan Taylor   
2024_pred_PPG_ppr          13.854265       13.677485        13.504416   
2024_pred_PosRank_ppr              7               8                9   

                               9              10            11  \
Player                 James Cook  Isiah Pacheco  Alvin Kamara   
2024_pred_PPG_ppr        13.35865      12.973047     12.959225   
2024_pred_PosRank_ppr          10             11            12   

                                       12           13              14  \
Player                 Kenneth Walker III  Josh Jacobs  Saquon Barkley   
2024_pred_PPG_ppr               12.343936    12.186932       12.072549   
2024_pred_PosRank_ppr                  13           14              15   

                                  15             16         17  \
Player                 D'Andre Swift  Jaylen Warren  Joe Mixon   
2024_pred_PPG_ppr          11.929216       11.91587  11.818624   
2024_pred_PosRank_ppr             16             17         18   

                                    18            19                20  \
Player                 Keaton Mitchell  James Conner  David Montgomery   
2024_pred_PPG_ppr            11.773307     11.747854         11.629289   
2024_pred_PosRank_ppr               19            20                21   

                                       21            22             23  \
Player                 Brian Robinson Jr.  Tony Pollard  Chuba Hubbard   
2024_pred_PPG_ppr               11.581104     11.308443       10.97479   
2024_pred_PosRank_ppr                  22            23             24   

                                 24  
Player                 Tyjae Spears  
2024_pred_PPG_ppr         10.971582  
2024_pred_PosRank_ppr            25

## Volume
Using a player's previous volume, can we predict their volume in the upcoming season?

In [ ]:
# group by each player and shift the 'Touches_per_game' column by 1
rb['NextSeason_Touches_per_game'] = rb.groupby('Key')['Touches_per_game'].shift(-1)

# check
rb[rb['Player'] == 'Christian McCaffrey'][['Year', 'Player', 'Touches_per_game', 'NextSeason_Touches_per_game']]

Year               Player  Touches_per_game  \
24712  2017  Christian McCaffrey         12.312500   
25283  2018  Christian McCaffrey         20.437500   
25899  2019  Christian McCaffrey         25.312500   
26709  2020  Christian McCaffrey         25.333333   
27186  2021  Christian McCaffrey         19.428571   
27889  2022  Christian McCaffrey         19.411765   
28524  2023  Christian McCaffrey         21.187500   

       NextSeason_Touches_per_game  
24712                    20.437500  
25283                    25.312500  
25899                    25.333333  
26709                    19.428571  
27186                    19.411765  
27889                    21.187500  
28524                          NaN

In [ ]:
# drop null targets
corr_df = rb.dropna(subset=['NextSeason_Touches_per_game'])

In [ ]:
# get correlation with NextSeason_Touches_per_game
corr = corr_df.corr()[['NextSeason_Touches_per_game']].sort_values(by='NextSeason_Touches_per_game', ascending=False).T
corr

NextSeason_Touches_per_game  PPGTarget_standard  \
NextSeason_Touches_per_game                          1.0            0.916551   

                             PPGTarget_half-ppr  PPGTarget_ppr  PPGTarget_6  \
NextSeason_Touches_per_game            0.908371       0.893887      0.89382   

                             SeasonTarget_standard  SeasonTarget_half-ppr  \
NextSeason_Touches_per_game               0.845681               0.834509   

                             SeasonTarget_ppr  SeasonTarget_6  \
NextSeason_Touches_per_game          0.819729        0.819653   

                             Rush_Yds_per_game  Scrim_Yds_per_game  \
NextSeason_Touches_per_game           0.640801            0.622108   

                             Touches_per_game  Rush_Att_per_game  Rush_Yds  \
NextSeason_Touches_per_game          0.620966           0.620426  0.614907   

                             PPG_standard  Rush_Att  Scrim_Yds   Touches  \
NextSeason_Touches_per_game      0.603932   0.59545    0.59077  0.590674   

                             PPG_half-ppr  Points_standard  \
NextSeason_Touches_per_game      0.590019         0.581157   

                             VORP_standard_12tm  VORP_standard_12tm_3WR  \
NextSeason_Touches_per_game            0.578722                0.578722   

                             VORP_standard_10tm  VORP_standard_10tm_3WR  \
NextSeason_Touches_per_game            0.573752                0.573752   

                                PPG_6   PPG_ppr  Points_half-ppr  \
NextSeason_Touches_per_game  0.573238  0.573133         0.566801   

                             VORP_half-ppr_12tm  VORP_half-ppr_12tm_3WR  \
NextSeason_Touches_per_game            0.565574                0.565574   

                             VORP_half-ppr_10tm  VORP_half-ppr_10tm_3WR  \
NextSeason_Touches_per_game            0.562075                0.562075   

                             VORP_6_12tm_3WR  VORP_6_12tm  Points_6  \
NextSeason_Touches_per_game         0.550907     0.550907  0.550844   

                             VORP_ppr_12tm  VORP_ppr_12tm_3WR  Points_ppr  \
NextSeason_Touches_per_game       0.550809           0.550809     0.55075   

                             VORP_6_10tm  VORP_6_10tm_3WR  VORP_ppr_10tm  \
NextSeason_Touches_per_game     0.546775         0.546775       0.546617   

                             VORP_ppr_10tm_3WR   Rush_TD  Scrim_TD  \
NextSeason_Touches_per_game           0.546617  0.485794  0.476065   

                             Rush_TD_per_game  Scrim_TD_per_game  \
NextSeason_Touches_per_game          0.474118           0.463492   

                             games_started_pct        GS   ProBowl       Fmb  \
NextSeason_Touches_per_game           0.416056  0.415188  0.384977  0.325113   

                              FmbLost  ...  games_played_pct    Rec_TD  \
NextSeason_Touches_per_game  0.317942  ...          0.130825  0.121539   

                             Pass_Att  Rec_TD_per_game  Pass_Att_per_game  \
NextSeason_Touches_per_game  0.110723         0.103371           0.098327   

                              Pass_TD  Pass_Cmp   Rec_Y/R  Pass_Yds  \
NextSeason_Touches_per_game  0.092404   0.09139  0.090653  0.090216   

                             Pass_TD_per_game  Pass_Cmp_per_game  \
NextSeason_Touches_per_game          0.085395           0.083673   

                             Pass_Yds_per_game      Cmp%      Year  \
NextSeason_Touches_per_game            0.07951  0.075377  0.074423   

                             PPTOvrRank_ppr  PPTOvrRank_6  Pass_Y/A  \
NextSeason_Touches_per_game        0.069707      0.066701  0.063541   

                             PPTPosRank_ppr  PPTPosRank_6  \
NextSeason_Touches_per_game        0.061315      0.061221   

                             PPTOvrRank_half-ppr  Pass_Int  Pass_Int_per_game  \
NextSeason_Touches_per_game             0.043056  0.042793           0.038582   

                             PPT_standard  PPTPosRank_half-ppr

- Rushing yds, Scrimmage yards, and touches per game in the prior season have strong correlations (0.6+) with the player's volume in the following season.